File: lib/src/call/strategy/sticky_handler_strategy.dart
import 'package:flutter/material.dart';
import 'package:flutter_bridge/src/call/handler.dart';
import 'package:flutter_bridge/src/call/strategy/handler_strategy.dart';

///   author : liuduo
///   e-mail : liuduo@gyenno.com
///   time   : 2022/08/27
///   desc   : 处理粘性数据
///   version: 1.0
abstract class StickyHandlerStrategy<A> implements HandlerStrategy<A> {
  final Map<String, A> _stickyData = {};

  @mustCallSuper
  @override
  void addHandler(Handler<A> handler) {
    var data = _stickyData[handler.name];
    print("flutter_bridge StickyHandlerStrategy addHandler ${handler.name}  $data");
    if (data != null) {
      try {
        handler.onCall(data);
      } on Exception {}
    }
  }

  @mustCallSuper
  @override
  Future<dynamic> onCallStrategy(String name, bool sticky, A data) {
    print("flutter_bridge StickyHandlerStrategy onCallStrategy $name $sticky $data");
    if (sticky) {
      _stickyData[name] = data;
    }
    return onCallNoSticky(name, data);
  }

  dynamic onCallNoSticky(String name, A data);
}

File: android/src/main/java/com/dodo/flutterbridge/call/strategy/StickyHandlerStrategy.kt
package com.dodo.flutterbridge.call.strategy

import androidx.annotation.CallSuper
import com.dodo.flutterbridge.call.Handler
import com.dodo.flutterbridge.call.exception.HandlerNotFoundException
import com.dodo.flutterbridge.call.exception.MutableHandlerException

/**
 *     author : liuduo
 *     e-mail : liuduo@gyenno.com
 *     time   : 2022/08/12
 *     desc   : 处理粘性数据
 *     version: 1.0
 */
abstract class StickyHandlerStrategy<A> : HandlerStrategy<A> {
    private val stickyData = mutableMapOf<String, A>()

    @CallSuper
    override fun addHandler(handler: Handler<A>) {
        val data = stickyData[handler.name]
        if (data != null) {
            handler.onCall(data)
        }
    }

    @CallSuper
    @Throws(HandlerNotFoundException::class, MutableHandlerException::class)
    override fun onCallStrategy(name: String, sticky: Boolean, data: A): Any? {
        if (sticky) {
            stickyData[name] = data
        }
        return onCallStrategy(name, data)
    }

    @Throws(HandlerNotFoundException::class, MutableHandlerException::class)
    abstract fun onCallStrategy(name: String, data: A): Any?
}